axiom/nr_attributes.c

```diff
@@ -162,10 +189,133 @@ static int nr_attribute_destination_modifier_compare(
   return 1;
 }
 
-void nr_attribute_config_modify_destinations(nr_attribute_config_t* config,
-                                             const char* match,
-                                             uint32_t include_destinations,
-                                             uint32_t exclude_destinations) {
+/*
+ * Purpose : Inspects current modifier list and adds finalize rules as needed.
+ *
+ * Params  : 1. Attribute configuration
+ *
+ * Notes   : Certain attriubtes like log context attributes expect the "include"
+ *           rules to act exclusively.
+ *           For example:
+ *             include = "A"
+ *             exclude = "B"
+ *             input = "A" "B" "C"
+ *             expected = "A"
+ *
+ *           Note that "C" was excluded because it was not in the include rules.
+ *           Also note an empty include rule means include everything and
+ *           and to exclude nothing.
+ *
+ *           All other attributes besides log context attributes do NOT have
+ *           this exclusive behavior for the include rules.  This function only
+ *           considers rules of the NR_ATTRIBUTE_DESTINATION_LOG destination.
+ *
+ *           The algorithm examines all input rules for the
+ *           NR_ATTRIBUTE_DESTINATION_LOG destination, and will create a new
+ *           rule of "exclude=*" with "is_finalize_rule=true" if any input 
+ *           rules exist.  The exception is if there is an input rule of "*"
+ *           for destination NR_ATTRIBUTE_DESTINATION_LOG.  In this case no 
+ *           finalize rule is added.
+ *
+ *           The net resulting effect is the exclusion of any attributes not
+ *           contained in the set of input rules for
+ *           NR_ATTRIBUTE_DESTINATION_LOG. The exception for if an "input=*"
+ *           rule is necessary since this input rule excludes nothing, making
+ *           the finalize rule of "exclude=*" unnecessary.
+ *
+ */
+static void nr_attribute_config_finalize_log_destination(
+    nr_attribute_config_t* config) {
+  nr_attribute_destination_modifier_t* cur = NULL;
+  nr_attribute_destination_modifier_t* prev = NULL;
+  nr_attribute_destination_modifier_t* next = NULL;
+  bool add_finalize_rule = false;
+
+  if (NULL == config || NULL == config->modifier_list) {
+    /* Since there is no configuration, no work to do. */
+    return;
+  }
+
+  /* remove any existing rules with is_finalize_rule = true */
+  for (cur = config->modifier_list; cur;) {
```

Comment:
Typically linked list traversal is done via a while loop rather than a for loop. I would suggest the following instead for clarity of purpose and keeping with implicit techniques, but the for loop method will work the same regardless:
```suggestion
  cur = config->modifier_list;
  while (NULL != cur) {
```